src/core/SampleBufferCache.cpp

```diff
@@ -0,0 +1,80 @@
+/*
+ * SampleBufferCache.cpp - Cache for storing SampleBuffer's
+ *
+ * Copyright (c) 2024 saker <sakertooth@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleBufferCache.h"
+
+#include <cassert>
+#include <cerrno>
+#include <stdexcept>
+
+namespace lmms {
+
+SampleBufferCache::SampleBufferCache(size_t capacity)
+	: m_capacity(capacity)
+{
+	if (capacity == 0) { throw std::runtime_error{"Capacity of cache cannot be zero"}; }
+}
+
+auto SampleBufferCache::add(const QString& key, Buffer buffer) -> bool
+{
+	if (buffer == nullptr || m_cache.contains(key)) { return false; }
+
+	assert(m_cache.size() <= m_capacity);
+	if (m_cache.size() == m_capacity)
+	{
+		const auto leastRecentlyUsed = std::min_element(m_cache.begin(), m_cache.end(),
+			[](const auto& entryOne, const auto& entryTwo) { return entryOne.use_count() < entryTwo.use_count(); });
+		m_cache.erase(leastRecentlyUsed);
```

Comment:
```suggestion
		const auto leastUsed = std::min_element(m_cache.begin(), m_cache.end(),
			[](const auto& entryOne, const auto& entryTwo) { return entryOne.use_count() < entryTwo.use_count(); });
		m_cache.erase(leastUsed);
```
Wouldn't this be more accurate name since it's just comparing how many times it is used?

Though I'm not sure how well a least-used cache would perform. To have a least-recently-used cache, you may need to change the type of `m_cache` to something like:
```c++
QHash<QString, std::pair<Buffer, std::chrono::steady_clock::time_point>> m_cache;
```
And update the time every time the buffer is added or retrieved from the cache.

Also, when a buffer is erased from the cache, maybe you should also remove its path from the file system watcher. Same with `SampleBufferCache::remove()`.

Reply:
I think we should just evict the ones that:
1. Aren't used at all
2. Been unused by anything for some time

Reply:
LRU by design will evict unused when another is loaded. We don't need to implement anything for 2.
Aren't used at all is harder to detect, but in essence is the same as 2, and will be evicted as the cache fills.

I'm debating whether the cache should even be an LRU instead of just an actual buffer cache (single load, multiple references).

Reply:
> I'm debating whether the cache should even be an LRU instead of just an actual buffer cache (single load, multiple references).

This is what I was trying to say. I dont think LRU is a good eviction policy for this problem. I think it makes more sense just to evict ones that are not used by anyone at all and its been that way for some time (maybe a minute), probably using some timer.

Reply:
No don't use a timer, that's making the problem more complex. Evict when the reference count is 0 as then it's no longer used, `shared_ptr`-like.

Reply:
> No don't use a timer, that's making the problem more complex. Evict when the reference count is 0 as then it's no longer used, shared_ptr-like.

But when should this eviction happen? Cleaning up the cache periodically simply resembles garbage collection, so I don't think this will drive up complexity at all.

Cleaning up the cache periodically would ensure that the buffers are guaranteed to be removed at some point once no one is using them. If we only do after some capacity that was set initially, a buffer can be added, never used again, and no one decides to max out the cache by adding more items, so it will just be there for nothing.

Removing them right after no one is using them doesn't help either because it may become in use again very soon, so we should allow some time to pass first.

Reply:
> But when should this eviction happen?
> Cleaning up the cache periodically would ensure that the buffers are guaranteed to be removed at some point once no one is using them.

If it's reference counted it will clean itself upon having 0 references, correct? We know when someone is using it because the counter is >0.

Reply:
> If it's reference counted it will clean itself upon having 0 references, correct? We know when someone is using it because the counter is >0.

Well yeah, I can add a custom deleter to the `shared_ptr` that will not only delete the buffer but also remove itself from the cache, but we forfeit any benefits if that sample will be accessed again after deletion without any changes being made on the filesystem.

Question becomes are we okay with that, and is it really a cache at that point?

Reply:
Maybe you could use the `std::shared_ptr` deleter to postpone the deletion of buffer it it is able to, else just delete it immediately.

Pseudocode:
```c++
// In the Entry class
struct BufferDeleter
{
	void operator()(SampleBuffer* ptr) const
	{
		if (!ptr) { return; }

		struct KeepAlive
		{
			std::shared_ptr<SampleBuffer> buffer;
		};

		auto temp = KeepAlive{ptr};

		// Now attempt to overwrite the Cache's std::weak_ptr value
		// for the filePath() key with `temp.buffer`.
		// While `temp` still exists, it will have sole
		// ownership of the buffer and the Cache will still be able
		// to use the buffer normally, but if time runs out and it is deleted,
		// the Cache will fail to lock the weak_ptr.

		if (/* if failed to overwrite cache */) { return; /* deletes buffer immediately */ }

		// Postpone buffer deletion (I don't know if this would work, but you get the idea)
		QTimer::singleShot(3000, [t = std::move(temp)](){});
	}
};

std::pair<QString, std::shared_ptr<const SampleBuffer, BufferDeleter>> m_entry;
```

Reply:
Sorry, I was overthinking it. The cache can store `weak_ptr`'s to the buffer, but return `shared_ptr`'s on calls to `SampleCache::get`. When no more `shared_ptr`'s are using that buffer, the `weak_ptr` in the cache will expire. If someone where to request that sample again and the `weak_ptr` has expired, its a cache miss and we have to reload the buffer back in. It's also a cache miss if the entry just does not exist in the cache. Otherwise, its a cache hit.

This is all I have for this class now. I think it works nicely:
```cpp
SampleCache::SampleCache()
{
	QObject::connect(&s_watcher, &QFileSystemWatcher::fileChanged, [](const QString& path) { evict(path); });
}

auto SampleCache::get(const QString& filePath) -> std::shared_ptr<const SampleBuffer>
{
	if (s_entries.contains(filePath) && !s_entries[filePath].expired()) { return s_entries[filePath].lock(); }
	s_entries[filePath] = std::make_shared<const SampleBuffer>(filePath);
	return s_entries[filePath].lock();
}

void SampleCache::evict(const QString& filePath)
{
	s_entries.remove(filePath);
	s_watcher.removePath(filePath);
}
```

Reply:
Perhaps we should move the QFileSystemWatcher to the GUI side and connect it to the cache there?

Reply:
> Perhaps we should move the QFileSystemWatcher to the GUI side and connect it to the cache there?

My initial idea was that the watcher would be separate from the cache, but @messmerd believed it should be together because you shouldn't forget to add the path to the watcher once you add it to the cache, which I agree with.

`QFileSystemWatcher` I dont think interferes with the core and GUI split. If you think about it, we can always replace the functionality of `QFileSystemWatcher` with another library that by no means is realted to the GUI. This is mostly functionality with watching changes on the filesystem, so it has nothing to do with the GUI.

But in terms of not using Qt in the core, it could be a problem, but not as much because in this case the functionality is so easily replaced with another standalone filesystem watcher library, so we can always change it when we are ready and it wont be a major blocking Qt dependency.

Reply:
@sakertooth 

Doesn't the code you posted delete buffers immediately once they're no longer used?
You said before:
> Removing them right after no one is using them doesn't help either because it may become in use again very soon, so we should allow some time to pass first.

Also, the `SampleCache::get()` method from the code you posted has a race condition with its `std::weak_ptr::expired()` usage, it searches `s_entries` twice rather than once, and I think it has some invalid behavior when assigning a `std::shared_ptr` temporary with sole ownership to the non-owning `std::weak_ptr`.

Here's something potentially better, though completely untested:
```c++
auto SampleCache::get(const QString& filePath) -> std::shared_ptr<const SampleBuffer>
{
	if (auto it = s_entries.find(filePath); it != s_entries.end())
	{
		if (auto buffer = it->lock()) { return buffer; }
	}

	auto buffer = std::make_shared<const SampleBuffer>(filePath);
	s_entries[filePath] = buffer;
	return buffer;
}
```

P.S. I will probably end up writing my own replacement for `QFileSystemWatcher` eventually since the CLAP host will need it.

Reply:
> Doesn't the code you posted delete buffers immediately once they're no longer used?

I'm debating on potential implementations. Do we want to evict buffers immediately when they are not used? This is what @Veratil was originally thinking we should do, and I'm not entirely against it. I only mentioned that we would forgo any retrieval speed benefits if that same sample will be feteched again in a short amount of time after eviction, but we still get the benefits of:

1. Fast load times if the buffer is being used and is within the cache
2. Memory savings if the same conditions are met in point 1

So it is just a question of if we want to forgo that benefit in the intrest of making the implementation simpler.

> Also, the `SampleCache::get()` method from the code you posted has a race condition with its `std::weak_ptr::expired()` usage, it searches `s_entries` twice rather than once, and I think it has some invalid behavior when assigning a `std::shared_ptr` temporary with sole ownership to the non-owning `std::weak_ptr`.

To be honest, I was not considering thread safety into this because I could only ever envision the GUI thread requesting samples from the cache. What other thread? When the user wants a sample, it must be from the GUI, no?

And for assigning an r-value `shared_ptr` to ` weak_ptr`, yeah. I originally stored it as an l-value and returned it directly (while also assigning it to the `weak_ptr`) but I couldn't see what was wrong with that at first.

Reply:
I understand your concerns better @messmerd. It actually does search twice, but its `O(2)` and not `O(n^2)` so its not a major concern, but still a useful optimization. Also, you're right that saying `!s_entries[filePath].expired()` is wrong assuming there are multiple threads because the `weak_ptr` may become expired right after we decided it wasn't, but it would only be a concern if multiple threads try and share the buffer, which I don't see happening?

I'll still apply the optimizations/fixes tough, along with some other things. I also decided that we should evict buffers immediately when they are no longer in use because the benefits we get from keeping that buffer for a minute or so isn't worth making the implementation more intricate. I'm still willing to negotiate different ideas though, but the two main benefits I listed above I feel is enough.